File: frontend/src/services/ParentStation/SessionListService/SessionList.ts
import { Map, List } from "immutable";
import moment from "moment";
import eventstore from "../../../eventstore";
import { Session } from "./types";

const WebSocketCloseCodeNormalClosure = 1000;
const WebSocketCloseCodeGoingAway = 1001;
// const WebSocketCloseCodeProtocolError = 1002;
// const WebSocketCloseCodeUnsupportedData = 1003;
// const WebSocketCloseCodeNoStatus = 1005;
// const WebSocketCloseCodeAbnormalClosure = 1006;

const expected_web_socket_close_codes = [
    WebSocketCloseCodeNormalClosure, // Manually closed by the user
    WebSocketCloseCodeGoingAway,     // Automatically sent by the browser when the tab is closed
];

interface SessionStartedEventData {
    id: string;
    host_connection_id: string;
    name: string
    started_at: string;
}

interface SessionStartedEvent extends eventstore.Event<SessionStartedEventData> {
    type: 'session.started';
}

interface SessionEndedEventData {
    id: string;
}

interface SessionEndedEvent extends eventstore.Event<SessionEndedEventData> {
    type: 'session.ended';
}

interface ClientConnectedEventData {
    client_id: string;
    connection_id: string;
    request_id: string;
}

interface ClientConnectedEvent extends eventstore.Event<ClientConnectedEventData> {
    type: 'client.connected';
}

interface ClientDisconnectedEventData {
    client_id: string;
    connection_id: string;
    request_id: string;
    web_socket_close_code: number;
}

interface ClientDisconnectedEvent extends eventstore.Event<ClientDisconnectedEventData> {
    type: 'client.disconnected';
}

interface ServerStartedEvent extends eventstore.Event<null> {
}

class SessionList {
    private sessionByID: Map<string, Session> = Map();
    private sessionByConnectionID: Map<string, Session> = Map();

    public seed = (sessions: Session[]) => {
        sessions.forEach(this.set_session);
    }

    public get_session_list = (): List<Session> => {
        return this.sessionByID.toList();
    }

    public apply = (event: eventstore.Event): boolean => {
        switch (event.type) {
            case 'session.started':
                return this.apply_session_started_event(event as SessionStartedEvent);
            case 'session.ended':
                return this.apply_session_ended_event(event as SessionEndedEvent);
            case 'client.connected':
                return this.apply_client_connected_event(event as ClientConnectedEvent);
            case 'client.disconnected':
                return this.apply_client_disconnected_event(event as ClientDisconnectedEvent);
            case 'server.started':
                return this.apply_server_started_event(event as ServerStartedEvent);
            default:
                return false;
        }
    }

    private set_session = (session: Session) => {
        this.sessionByID = this.sessionByID.set(session.id, session);
        this.sessionByConnectionID = this.sessionByConnectionID.set(session.host_connection_id, session);
    }

    private delete_session = (session: Session) => {
        this.sessionByID = this.sessionByID.delete(session.id);
        this.sessionByConnectionID = this.sessionByConnectionID.delete(session.host_connection_id);
    }

    private apply_session_started_event = (event: SessionStartedEvent): boolean => {
        const session: Session = {
            id: event.data.id,
            name: event.data.name,
            host_connection_id: event.data.host_connection_id,
            started_at: moment(event.data.started_at, eventstore.MomentFormatRFC3339),
            host_connection_state: {
                state: 'connected',
                request_id: 'TODO', // TODO
                since: moment(event.unix_timestamp, eventstore.MomentFormatUnixTimestamp), // TODO this is the session started timestamp, not the host connected timestamp
            },
        };
        const existing_session = this.sessionByConnectionID.get(session.host_connection_id);
        if (existing_session !== undefined)
            this.delete_session(existing_session);
        this.set_session(session);
        return true;
    }

    private apply_session_ended_event = (event: SessionEndedEvent): boolean => {
        const session = this.sessionByID.get(event.data.id);
        if (session === undefined) return false;
        this.delete_session(session);
        return true;
    }

    private apply_client_connected_event = (event: ClientConnectedEvent): boolean => {
        const session = this.sessionByConnectionID.get(event.data.connection_id);
        if (session === undefined) return false;
        const updated_session: Session = {
            ...session,
            host_connection_state: {
                state: 'connected',
                request_id: event.data.request_id,
                since: moment(event.unix_timestamp, eventstore.MomentFormatUnixTimestamp),
            },
        };
        this.set_session(updated_session);
        return true;
    }

    private apply_client_disconnected_event = (event: ClientDisconnectedEvent): boolean => {
        const session = this.sessionByConnectionID.get(event.data.connection_id);
        if (session === undefined) return false;
        if (session.host_connection_state.state !== 'connected') return false;
        if (session.host_connection_state.request_id !== event.data.request_id) return false;
        const expected_web_socket_close_code_received = expected_web_socket_close_codes.includes(event.data.web_socket_close_code);
        if (!expected_web_socket_close_code_received) {
            const updated_session: Session = {
                ...session,
                host_connection_state: {
                    state: 'disconnected',
                    since: moment(event.unix_timestamp, eventstore.MomentFormatUnixTimestamp),
                },
            };
            this.set_session(updated_session);
            return true;
        }
        this.delete_session(session);
        return true;
    }

    private apply_server_started_event = (event: ServerStartedEvent): boolean => {
        if (this.sessionByConnectionID.isEmpty()) return false;
        const since = moment(event.unix_timestamp, eventstore.MomentFormatUnixTimestamp);
        this.sessionByConnectionID.forEach((session) => {
            if (session.host_connection_state.state === 'disconnected') return;
            const updated_session: Session = {
                ...session,
                host_connection_state: {
                    state: 'disconnected',
                    since: since,
                },
            };
            this.set_session(updated_session);
        });
        return true;
    }
}

export default SessionList;

File: golang/internal/sessionlist/SessionList.go
package sessionlist

import (
	"context"
	"encoding/json"
	"sort"
	"sync"
	"time"

	"github.com/Ryan-A-B/beddybytes/golang/internal/contextx"
	"github.com/Ryan-A-B/beddybytes/golang/internal/eventlog"
	"github.com/Ryan-A-B/beddybytes/golang/internal/fatal"
	"github.com/gorilla/websocket"
)

type SessionList struct {
	mutex    sync.Mutex
	log      eventlog.EventLog
	cursor   int64
	sessions []*Session
}

type NewInput struct {
	Log eventlog.EventLog
}

func New(ctx context.Context, input NewInput) *SessionList {
	sessionList := &SessionList{
		log: input.Log,
	}
	go sessionList.catchUp(ctx)
	return sessionList
}

type ListOutput struct {
	Cursor   int64      `json:"cursor"`
	Sessions []*Session `json:"sessions"`
}

func (sessionList *SessionList) List(ctx context.Context) (output ListOutput) {
	sessionList.catchUp(ctx)
	accountID := contextx.GetAccountID(ctx)
	index := sort.Search(len(sessionList.sessions), func(i int) bool {
		return sessionList.sessions[i].AccountID >= accountID
	})
	output.Cursor = sessionList.cursor
	output.Sessions = make([]*Session, 0, len(sessionList.sessions)-index)
	for i := index; i < len(sessionList.sessions); i++ {
		session := sessionList.sessions[i]
		if session.AccountID != accountID {
			return
		}
		output.Sessions = append(output.Sessions, session)
	}
	return
}

func (sessionList *SessionList) catchUp(ctx context.Context) {
	sessionList.mutex.Lock()
	defer sessionList.mutex.Unlock()
	iterator := sessionList.log.GetEventIterator(ctx, eventlog.GetEventIteratorInput{
		FromCursor: sessionList.cursor,
	})
	for iterator.Next(ctx) {
		event := iterator.Event()
		sessionList.apply(ctx, event)
		sessionList.cursor = event.LogicalClock
	}
	fatal.OnError(iterator.Err())
}

func (sessionList *SessionList) apply(ctx context.Context, event *eventlog.Event) {
	apply, ok := applyByType[event.Type]
	if !ok {
		return
	}
	apply(ctx, sessionList, event)
}

func (sessionList *SessionList) put(session *Session) {
	index := sessionList.search(session.AccountID, session.ID)
	if index == len(sessionList.sessions) {
		sessionList.sessions = append(sessionList.sessions, session)
		return
	}
	if sessionList.sessions[index].ID == session.ID {
		sessionList.sessions[index] = session
		return
	}
	sessionList.sessions = append(sessionList.sessions, nil)
	copy(sessionList.sessions[index+1:], sessionList.sessions[index:])
	sessionList.sessions[index] = session
}

func (sessionList *SessionList) delete(accountID string, sessionID string) {
	index := sessionList.search(accountID, sessionID)
	if index == len(sessionList.sessions) {
		return
	}
	session := sessionList.sessions[index]
	if session.AccountID != accountID || session.ID != sessionID {
		return
	}
	sessionList.sessions = append(sessionList.sessions[:index], sessionList.sessions[index+1:]...)
}

func (sessionList *SessionList) search(accountID string, sessionID string) int {
	return sort.Search(len(sessionList.sessions), func(i int) bool {
		if sessionList.sessions[i].AccountID > accountID {
			return true
		}
		if sessionList.sessions[i].AccountID != accountID {
			return false
		}
		return sessionList.sessions[i].ID >= sessionID
	})
}

func (sessionList *SessionList) getSessionByConnectionID(accountID string, connectionID string) (session *Session, ok bool) {
	index := sort.Search(len(sessionList.sessions), func(i int) bool {
		return sessionList.sessions[i].AccountID >= accountID
	})
	for i := index; i < len(sessionList.sessions); i++ {
		session = sessionList.sessions[i]
		if session.AccountID != accountID {
			return
		}
		if session.HostConnectionID == connectionID {
			ok = true
			return
		}
	}
	return
}

type applyFunc func(ctx context.Context, sessionList *SessionList, event *eventlog.Event)

const EventTypeServerStarted = "server.started"
const EventTypeSessionStarted = "session.started"
const EventTypeSessionEnded = "session.ended"
const EventTypeClientConnected = "client.connected"
const EventTypeClientDisconnected = "client.disconnected"

var applyByType = map[string]applyFunc{
	EventTypeServerStarted:      applyServerStartedEvent,
	EventTypeSessionStarted:     applySessionStartedEvent,
	EventTypeSessionEnded:       applySessionEndedEvent,
	EventTypeClientConnected:    applyClientConnectedEvent,
	EventTypeClientDisconnected: applyClientDisconnectedEvent,
}

type SessionStartedEventData struct {
	ID               string    `json:"id"`
	Name             string    `json:"name"`
	HostConnectionID string    `json:"host_connection_id"`
	StartedAt        time.Time `json:"started_at"`
}

func applySessionStartedEvent(ctx context.Context, sessionList *SessionList, event *eventlog.Event) {
	var sessionStartedEventData SessionStartedEventData
	err := json.Unmarshal(event.Data, &sessionStartedEventData)
	fatal.OnError(err)
	existingSession, ok := sessionList.getSessionByConnectionID(event.AccountID, sessionStartedEventData.HostConnectionID)
	if ok {
		sessionList.delete(event.AccountID, existingSession.ID)
	}
	sessionList.put(&Session{
		AccountID:        event.AccountID,
		ID:               sessionStartedEventData.ID,
		Name:             sessionStartedEventData.Name,
		HostConnectionID: sessionStartedEventData.HostConnectionID,
		StartedAt:        sessionStartedEventData.StartedAt,
		HostConnectionState: HostConnectionStateConnected{
			HostConnectionStateBase: HostConnectionStateBase{
				State: ConnectionStateConnected,
				Since: event.UnixTimestamp,
			},
			RequestID: "TODO", // TODO
		},
	})
}

type SessionEndedEventData struct {
	ID string `json:"id"`
}

func applySessionEndedEvent(ctx context.Context, sessionList *SessionList, event *eventlog.Event) {
	var sessionEndedEventData SessionEndedEventData
	err := json.Unmarshal(event.Data, &sessionEndedEventData)
	fatal.OnError(err)
	sessionList.delete(event.AccountID, sessionEndedEventData.ID)
}

type ClientDisconnectedEventData struct {
	ClientID           string `json:"client_id"`
	ConnectionID       string `json:"connection_id"`
	RequestID          string `json:"request_id"`
	WebSocketCloseCode int    `json:"web_socket_close_code"`
}

func applyClientDisconnectedEvent(ctx context.Context, sessionList *SessionList, event *eventlog.Event) {
	var clientDisconnectedEventData ClientDisconnectedEventData
	err := json.Unmarshal(event.Data, &clientDisconnectedEventData)
	fatal.OnError(err)
	session, ok := sessionList.getSessionByConnectionID(event.AccountID, clientDisconnectedEventData.ConnectionID)
	if !ok {
		return
	}
	if session.HostConnectionState.GetState() != ConnectionStateConnected {
		return
	}
	hostConnectionState := session.HostConnectionState.(HostConnectionStateConnected)
	if hostConnectionState.RequestID != clientDisconnectedEventData.RequestID {
		return
	}
	switch clientDisconnectedEventData.WebSocketCloseCode {
	case websocket.CloseNormalClosure, websocket.CloseGoingAway:
		sessionList.delete(event.AccountID, session.ID)
	default:
		session.HostConnectionState = HostConnectionStateDisconnected{
			HostConnectionStateBase: HostConnectionStateBase{
				State: ConnectionStateDisconnected,
				Since: event.UnixTimestamp,
			},
		}
	}
}

type ClientConnectedEventData struct {
	ClientID     string `json:"client_id"`
	ConnectionID string `json:"connection_id"`
	RequestID    string `json:"request_id"`
}

func applyClientConnectedEvent(ctx context.Context, sessionList *SessionList, event *eventlog.Event) {
	var clientConnectedEventData ClientConnectedEventData
	err := json.Unmarshal(event.Data, &clientConnectedEventData)
	fatal.OnError(err)
	session, ok := sessionList.getSessionByConnectionID(event.AccountID, clientConnectedEventData.ConnectionID)
	if !ok {
		return
	}
	session.HostConnectionState = HostConnectionStateConnected{
		HostConnectionStateBase: HostConnectionStateBase{
			State: ConnectionStateConnected,
			Since: event.UnixTimestamp,
		},
		RequestID: clientConnectedEventData.RequestID,
	}
}

func applyServerStartedEvent(ctx context.Context, sessionList *SessionList, event *eventlog.Event) {
	for _, session := range sessionList.sessions {
		if session.HostConnectionState.GetState() == ConnectionStateDisconnected {
			continue
		}
		session.HostConnectionState = HostConnectionStateDisconnected{
			HostConnectionStateBase: HostConnectionStateBase{
				State: ConnectionStateDisconnected,
				Since: event.UnixTimestamp,
			},
		}
	}
}
